src/mpid/ch4/shm/xpmem/xpmem_init.c

```diff
@@ -92,10 +119,23 @@ int MPIDI_XPMEM_mpi_finalize_hook(void)
     MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPIDI_XPMEM_FINALIZE_HOOK);
     MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPIDI_XPMEM_FINALIZE_HOOK);
 
+    /* Free the rest of counter obj */
+    while (MPIDI_XPMEM_global.num_pending_cnt)
+        MPIDI_Progress_test(MPIDI_PROGRESS_SHM);
```

Comment:
The `num_pending_cnt ` has to be an atomic counter (`MPIR_cc_t`) because it might be updated by different threads.

Reply:
Can you please make the comment a little more detailed? How about "Ensure all counter objs are freed at MPIDI_XPMEM_ctrl_send_lmt_cnt_free_cb"?